### PR TITLE
moved graveyard generation out of reactor

### PR DESCRIPTION
**Pull request by Shimwell**
## Proposed changes

This PR moves some of the graveyard creation from Reactor

the largest_dimension has been moved into the attributes of of Shape objects
the 3d hollow cube creation has been moved into a new parametric component called HollowCube.

This is useful as the the Shape objects will need their own Graveyard bounding box if we want to do neutronics simulations on the Shapes.

@billingsley-john  I think the inclusion of a Dagmc Graveyard for Shape objects will now be easier as the useful code is now in more accessable places in the code

This solves #586

Also the largest_dimention attribute can be used to determine 3D and 2D mesh tallies sizes in the neutronics simulations

## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] Pep8 applied
- [ ] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: paramak/parametric_components/hollow_cube.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+
+import cadquery as cq
+from paramak import RotateStraightShape
+from paramak.utils import get_hash
+
+
+class HollowCube(RotateStraightShape):
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/hollow_cube.py:7]: As far as I understood, this doesn't use any of the methods of `RotateStraightShape`. Shouldn't it simply inherit from `Shape` ?

File: paramak/parametric_components/hollow_cube.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    """A hollow cube with a constant thickness. Can be used to create a DAGMC
+    Graveyard.
+
+    Arguments:
+        length (float): The length to use for the height, width, depth of the
+            inner dimentions of the cube.
+        thickness (float): thickness of the vessel
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/hollow_cube.py:14]:
```suggestion
        thickness (float, optional): thickness of the vessel. Defaults to 10.0.
```

File: paramak/parametric_components/hollow_cube.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            self.length + self.thickness
+        )
+
+        # subtracts the two boxes to leave a hollow box
+        new_shape = outer_box.cut(inner_box)
+
+        self.solid = new_shape
+
+        # Calculate hash value for current solid
+        self.hash_value = get_hash(self)
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/hollow_cube.py:76]: Not needed here cause the hash_value is set in Shape().solid getter.

**Review by RemDelaporteMathurin — COMMENTED** [review]
Great stuff @Shimwell ! Left a few comments. Moreover, it seems like additionnal tests are adding cause the diff coverage isn't that good. See in Files changed the lines that aren't hit by the test suite.

File: paramak/parametric_components/hollow_cube.py (hunk 1 above)
Comment (Shimwell) [inline paramak/parametric_components/hollow_cube.py:7]:
```suggestion
class HollowCube(Shape):
```

File: paramak/shape.py [hunk 4]
```diff
@@ -158,6 +158,39 @@ def union(self):
...
+            largest_dimension = max(
+                abs(self.solid.val().BoundingBox().xmax),
+                abs(self.solid.val().BoundingBox().xmin),
+                abs(self.solid.val().BoundingBox().ymax),
+                abs(self.solid.val().BoundingBox().ymin),
+                abs(self.solid.val().BoundingBox().zmax),
+                abs(self.solid.val().BoundingBox().zmin),
+                largest_dimension
+            )
+        self._largest_dimension = largest_dimension
```
Comment (RemDelaporteMathurin) [inline paramak/shape.py:187]:
```suggestion
        self.largest_dimension = largest_dimension
```
If underscore is used, `@largest_dimension.setter` won't be used.

File: paramak/parametric_components/hollow_cube.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+
+import cadquery as cq
+from paramak import RotateStraightShape
```
Comment (RemDelaporteMathurin) [inline paramak/parametric_components/hollow_cube.py:3]:
```suggestion
from paramak import Shape
```

File: paramak/parametric_components/hollow_cube.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+
+import cadquery as cq
+from paramak import Shape
+from paramak.utils import get_hash
```
Comment (Shimwell) [inline paramak/parametric_components/hollow_cube.py:4]:
```suggestion
```

**Review by Shimwell — COMMENTED** [review]
looks fine now, thanks for the tidy up Remi

**Pull request merged**